src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -1529,5 +1530,218 @@ private bool TryGetNumberLookForEnd(ReadOnlySpan<byte> data, out ReadOnlySpan<by
+            return default;
+        }
+
+        public float GetValueAsSingle()
+        {
+            //TODO: We know whether this is true or not ahead of time
+            if (Value.IndexOfAny((byte)'e', (byte)'E') == -1)
+            {
+                if (Utf8Parser.TryParse(Value, out float value, out int bytesConsumed))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            else
+            {
+                if (Utf8Parser.TryParse(Value, out float value, out int bytesConsumed, 'e'))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            ThrowInvalidCastException();
+            return default;
+        }
+
+        public double GetValueAsDouble()
+        {
+            if (Value.IndexOfAny((byte)'e', (byte)'E') == -1)
+            {
+                if (Utf8Parser.TryParse(Value, out double value, out int bytesConsumed))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            else
+            {
+                if (Utf8Parser.TryParse(Value, out double value, out int bytesConsumed, standardFormat: 'e'))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            ThrowInvalidCastException();
+            return default;
+        }
+
+        public decimal GetValueAsDecimal()
+        {
+            if (Value.IndexOfAny((byte)'e', (byte)'E') == -1)
+            {
+                if (Utf8Parser.TryParse(Value, out decimal value, out int bytesConsumed))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            else
+            {
+                if (Utf8Parser.TryParse(Value, out decimal value, out int bytesConsumed, standardFormat: 'e'))
+                {
+                    if (Value.Length == bytesConsumed)
+                    {
+                        return value;
+                    }
+                }
+            }
+            ThrowInvalidCastException();
+            return default;
+        }
+
+        public object GetValueAsNumber()
+        {
+            if (Utf8Parser.TryParse(Value, out int intVal, out int bytesConsumed))
+            {
+                if (Value.Length == bytesConsumed)
+                {
+                    return intVal;
+                }
+            }
+
+            if (Utf8Parser.TryParse(Value, out long longVal, out bytesConsumed))
```

Comment:
should this method first try to parse as a long and then iff the value fits in an int, cast and return?

Reply:
The `TryParse(out long)` API is ~10% slower than the one that returns an int. It is also more likely for the input to contain an int.